File: Python/Coupon.py
class Coupon:
	"Coupon class"
	def __init__(self, name, rate):
		self.name = name
		self.rate = rate

	def getName(self):
		return self.name

	def getRate(self):
		return self.rate

	def payment(self, price):
		"Take in a price and return the paid sum"
		pay = 0.0
		if self.rate < 1: # discount coupon
			pay = (1-self.rate) * price
		else: # cash coupon
			if price > self.rate: # price larger value over coupon
				pay = self.rate - price
			else: # vice versa
				pay = price - self.rate
		return pay

price, num_coupon = raw_input().split()
price = float(price)
num_coupon = int(num_coupon)
coupon = [None] * num_coupon
pay = []
least_theory_pay = []

for i in range(num_coupon):
	name,rate = raw_input().split()
	rate = float(rate)
	coupon[i] = Coupon(name, rate)
	pay.append(coupon[i].payment(price))
	least_theory_pay.append(abs(coupon[i].payment(price)))
value_pay = least_theory_pay[0]
for i in range(1, num_coupon):
	if least_theory_pay[i] < value_pay:
		value_pay = least_theory_pay[i]
		index = i

if pay[index] < 0:
	value_pay = 0.0
print "Best choice:", coupon[index].getName()
print "You need to pay $%.2f" % value_pay

File: JAVA/Coupon.java
import java.util.Scanner;
import java.util.Arrays;
import java.lang.Math;

class Coupon {
	//Attributes
	private String name;
	private double rate;

	//Constructors
	public Coupon() {}
	public Coupon(String new_name, double new_rate) {
		this.name = new_name;
		this.rate = new_rate;
	}

	//Accessors 
	public String getName() {
		return this.name;
	}
	public double getRate() {
		return this.rate;
	}
	//Methods
	public double payment(double price) {
		double pay =0.0;
		if (this.rate < 1) { //discount coupon
			pay = (1-rate) * price;
		} else { //cash coupon
			if (price > this.rate) {
				pay = this.rate - price;
			} else {
				pay = price - this.rate;
			}
		}
		return pay;
	}

}

class Redeem {
	public static void main(String[] args) {
		//Declare a scanner object to read input
		Scanner scan = new Scanner(System.in);
		//Declare the necessary variables
		double price, rate, value_pay; 
		int index=0, num_coupon;
		String name;
		//Read input and process them accordingly
		price = scan.nextDouble();
		num_coupon = scan.nextInt();
		Coupon[] coupon = new Coupon[num_coupon];
		double[] pay = new double[num_coupon];
		double[] theory_pay = new double[num_coupon];

		for (int i=0; i < num_coupon; i++) {// Instantiate coupons
			name = scan.next();
			rate = scan.nextDouble();
			coupon[i] = new Coupon(name, rate);
			pay[i] = coupon[i].payment(price);
			theory_pay[i] = Math.abs(pay[i]);
		}
		value_pay = theory_pay[0];
		for (int i=0; i < num_coupon; i++) {
			if (theory_pay[i] < value_pay) {
				value_pay = theory_pay[i];
				index = i;
			}
		}
		if (pay[index] < 0) {
			value_pay = 0.0;
		}
		System.out.println("Best choice: " + coupon[index].getName());
		System.out.println("You need to pay $" + value_pay);
	}
}